File: internal/study/gate.go
package study

import "github.com/sebastiancaraballo/polyglot/internal/model"

// The Foundations decoding gate. Polyglot follows the Simple View of Reading:
// comprehension can only ride on top of decoding, so the learner must be able to
// read kana fluently before meeting Japanese words and sentences. For a learner
// whose native script is the Latin alphabet (Spanish L1), kana is entirely new
// decoding, so this gate stands at the very start of the course.
//
// The gate is measured over the *base gojūon* of each syllabary — the core set
// you must decode automatically. Dakuten, handakuten, and combination kana are
// still practiced and reviewed, but they are derived forms and do not hold the
// gate.

// Fluency summarizes how much of one syllabary's base set a learner has driven
// to automaticity.
type Fluency struct {
	Mastered int
	Total    int
}

// Fluent reports whether every base kana in the syllabary is mastered.
func (f Fluency) Fluent() bool { return f.Total > 0 && f.Mastered >= f.Total }

// kanaFluency computes fluency over the base gojūon of one syllabary.
func kanaFluency(items []model.KanaItem, progress map[string]model.KanaProgress, typ model.KanaType) Fluency {
	var f Fluency
	for _, it := range items {
		if it.Type != typ || it.Category != model.Base {
			continue
		}
		f.Total++
		if progress[it.Char].Mastered {
			f.Mastered++
		}
	}
	return f
}

// Gate is the Foundations decoding gate, derived from a learner's kana progress.
// Katakana practice unlocks once hiragana is fluent; reading Japanese words and
// sentences unlocks once both syllabaries are fluent.
type Gate struct {
	Hiragana Fluency
	Katakana Fluency
}

// NewGate computes the gate from the full kana set and a learner's progress.
func NewGate(items []model.KanaItem, progress map[string]model.KanaProgress) Gate {
	return Gate{
		Hiragana: kanaFluency(items, progress, model.Hiragana),
		Katakana: kanaFluency(items, progress, model.Katakana),
	}
}

// KatakanaUnlocked reports whether katakana practice is available yet.
func (g Gate) KatakanaUnlocked() bool { return g.Hiragana.Fluent() }

// ReadingUnlocked reports whether reading Japanese words and sentences is
// available yet: both syllabaries must be fluent.
func (g Gate) ReadingUnlocked() bool { return g.Hiragana.Fluent() && g.Katakana.Fluent() }

File: core/src/study/gate.rs
use std::collections::HashMap;

use crate::model::{KanaCategory, KanaItem, KanaProgress, KanaType};

/// Summarizes how much of one syllabary's base set a learner has driven to
/// automaticity.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct Fluency {
    pub mastered: i64,
    pub total: i64,
}

impl Fluency {
    /// Reports whether every base kana in the syllabary is mastered.
    pub fn fluent(&self) -> bool {
        self.total > 0 && self.mastered >= self.total
    }
}

fn kana_fluency(
    items: &[KanaItem],
    progress: &HashMap<String, KanaProgress>,
    typ: KanaType,
) -> Fluency {
    let mut f = Fluency::default();
    for it in items {
        if it.kana_type != typ || it.category != KanaCategory::Base {
            continue;
        }
        f.total += 1;
        if progress.get(&it.char).is_some_and(|p| p.mastered) {
            f.mastered += 1;
        }
    }
    f
}

/// The Foundations decoding gate, derived from a learner's kana progress.
/// Katakana practice unlocks once hiragana is fluent; reading Japanese words
/// and sentences unlocks once both syllabaries are fluent.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct Gate {
    pub hiragana: Fluency,
    pub katakana: Fluency,
}

impl Gate {
    /// Reports whether katakana practice is available yet.
    pub fn katakana_unlocked(&self) -> bool {
        self.hiragana.fluent()
    }

    /// Reports whether reading Japanese words and sentences is available yet:
    /// both syllabaries must be fluent.
    pub fn reading_unlocked(&self) -> bool {
        self.hiragana.fluent() && self.katakana.fluent()
    }
}

/// Computes the gate from the full kana set and a learner's progress.
pub fn new_gate(items: &[KanaItem], progress: &HashMap<String, KanaProgress>) -> Gate {
    Gate {
        hiragana: kana_fluency(items, progress, KanaType::Hiragana),
        katakana: kana_fluency(items, progress, KanaType::Katakana),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base(char: &str, typ: KanaType) -> KanaItem {
        KanaItem {
            char: char.to_string(),
            romaji: String::new(),
            kana_type: typ,
            category: KanaCategory::Base,
        }
    }

    fn mastered(chars: &[&str]) -> HashMap<String, KanaProgress> {
        chars
            .iter()
            .map(|c| {
                (
                    c.to_string(),
                    KanaProgress {
                        char: c.to_string(),
                        mastered: true,
                        ..Default::default()
                    },
                )
            })
            .collect()
    }

    #[test]
    fn gate_unlocks_progressively() {
        let items = vec![
            base("あ", KanaType::Hiragana),
            base("い", KanaType::Hiragana),
            base("ア", KanaType::Katakana),
        ];

        let g = new_gate(&items, &mastered(&[]));
        assert!(!g.katakana_unlocked());
        assert!(!g.reading_unlocked());

        let g = new_gate(&items, &mastered(&["あ", "い"]));
        assert!(g.katakana_unlocked(), "hiragana fluent unlocks katakana");
        assert!(!g.reading_unlocked());

        let g = new_gate(&items, &mastered(&["あ", "い", "ア"]));
        assert!(g.reading_unlocked(), "both fluent unlocks reading");
    }
}
